src/include/sof/audio/audio_stream.h

```diff
@@ -338,6 +369,22 @@ audio_stream_get_free_frames(const struct audio_stream *stream)
 		audio_stream_frame_bytes(stream);
 }
 
+/**
+ * Calculates free space in frames, which is aligned with stream->frame_align_shift-th
+ * power of 2, handling underrun_permitted behaviour
+ * @param stream Stream pointer
+ * @return amount of space free in frames
+ */
+static inline uint32_t
+audio_stream_get_free_frames_aligned(const struct audio_stream *stream)
+{
+	uint32_t frames = audio_stream_get_free_bytes(stream) /
+			  audio_stream_frame_bytes(stream);
```

Comment:
I really want to avoid this division - there must be a way (that I may not have described in previous comments to full detail) where we can 
1) Get the free/avail bytes.
2) Get the nearest frame alignment by shifting, adding or masking only.